src/core/thread/network_data_notifier.cpp

```diff
@@ -0,0 +1,185 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements transmissions of SVR_DATA.ntf messages to the Leader.
+ */
+
+#include "network_data_notifier.hpp"
+
+#if OPENTHREAD_FTD || OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE || OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator-getters.hpp"
+#include "thread/network_data_leader.hpp"
+#include "thread/network_data_local.hpp"
+
+namespace ot {
+namespace NetworkData {
+
+Notifier::Notifier(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mNotifierCallback(aInstance, &Notifier::HandleStateChanged, this)
+    , mTimer(aInstance, &Notifier::HandleTimer, this)
+    , mState(kStateNewServiceData)
+    , mWaitingForResponse(false)
+{
+}
+
+void Notifier::ServerDataUpdated(void)
+{
+    mState = kStateNewServiceData;
+    SynchronizeNetworkData();
+}
+
+void Notifier::SynchronizeNetworkData(void)
+{
+    otError error = OT_ERROR_NOT_FOUND;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsAttached() && !mWaitingForResponse);
+
+    VerifyOrExit((mState == kStateNewServiceData) || !mTimer.IsRunning());
+
+#if OPENTHREAD_FTD
+    mState = kStateRemoveStaleChildren;
+    error  = Get<Leader>().RemoveStaleChildEntries(&Notifier::HandleCoapResponse, this);
+    VerifyOrExit(error == OT_ERROR_NOT_FOUND);
+#endif
+
+#if OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE || OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
+    mState = kStateSynchronizeServerData;
+    error  = Get<Local>().UpdateInconsistentServerData(&Notifier::HandleCoapResponse, this);
+    VerifyOrExit(error == OT_ERROR_NOT_FOUND);
+#endif
+
+    mState = kStateSynchronized;
```

Comment:
I have a suggestion here for the states. Not entirely sure whether it would help simplify the code/logic. 

If I understand it correctly, the `mState` serves two purpose:
- We have `mState = kStateNewServiceData;` which indicates there is some pending change we need to do the sync (and we can enter this from any other state on `ServerDataUpdated()` call or `HandleStateChanged()` call).
- The other states are mainly used to decide the next timer interval from `HandleCoapResponse()` when `aResult` is `OT_ERROR_NONE`.

How about the following:
- We track the timer wait interval in `mNextDelay` which is set from `SynchronizeNetworkData()` (based on what we are doing e.g. when removing a stale child we set it to `kDelayRemoveStaleChildren`, when synchronizing  we set it to `kDelaySynchronizeServerData`). 
- From  `HandleCoapResponse()` we use the `mNextDelay` to restart the timer to do next round of `SynchronizeNetworkData()`.
- For  `mState = kStateNewServiceData` we can either use a new boolean variable to track or maybe we set the `mNextDelay` to zero to indicate this?

@jwhui , what do you think?